xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/DefaultLbStateFactory.java

```diff
@@ -0,0 +1,385 @@
+            perPriorityDegradedBuilder.put(priority, healthAndDegraded.degradedWeight);
+        }
+        PerPriorityLoad perPriorityLoad = buildLoads(prioritySet,
+                                                     perPriorityHealthBuilder.build(),
+                                                     perPriorityDegradedBuilder.build());
+        final PerPriorityPanic perPriorityPanic =
+                recalculatePerPriorityPanic(prioritySet,
+                                            perPriorityLoad.normalizedTotalAvailability());
+
+        logger.debug("XdsEndpointGroup load balancer priorities for cluster({}) has been updated with" +
+                     " perPriorityLoad({}), perPriorityPanic({}).",
+                     prioritySet.cluster().getName(), perPriorityLoad, perPriorityPanic);
+
+        if (perPriorityPanic.totalPanic()) {
+            perPriorityLoad = recalculateLoadInTotalPanic(prioritySet);
+            logger.debug("XdsEndpointGroup load balancer in panic for cluster({}) with perPriorityLoad({}).",
+                         prioritySet.cluster().getName(), perPriorityLoad);
+        }
+        return new DefaultLbState(prioritySet, perPriorityLoad, perPriorityPanic);
+    }
+
+    private static HealthAndDegraded recalculatePerPriorityState(
+            int priority, PrioritySet prioritySet) {
+        final HostSet hostSet = prioritySet.hostSets().get(priority);
+        final int hostCount = hostSet.hosts().size();
+
+        if (hostCount > 0) {
+            long healthyWeight = 0;
+            long degradedWeight = 0;
+            long totalWeight = 0;
+            if (hostSet.weightedPriorityHealth()) {
+                for (Endpoint host : hostSet.healthyHosts()) {
+                    healthyWeight += host.weight();
+                }
+                for (Endpoint host : hostSet.degradedHosts()) {
+                    degradedWeight += host.weight();
+                }
+                for (Endpoint host : hostSet.hosts()) {
+                    totalWeight += host.weight();
+                }
+            } else {
+                healthyWeight = hostSet.healthyHosts().size();
+                degradedWeight = hostSet.degradedHosts().size();
+                totalWeight = hostCount;
+            }
+            final int health = (int) Math.min(100L, LongMath.saturatedMultiply(
+                    hostSet.overProvisioningFactor(), healthyWeight) / totalWeight);
+            final int degraded = (int) Math.min(100L, LongMath.saturatedMultiply(
+                    hostSet.overProvisioningFactor(), degradedWeight) / totalWeight);
+            return new HealthAndDegraded(health, degraded);
+        }
+        return new HealthAndDegraded(0, 0);
+    }
+
+    private static PerPriorityLoad buildLoads(PrioritySet prioritySet,
+                                              Map<Integer, Integer> perPriorityHealth,
+                                              Map<Integer, Integer> perPriorityDegraded) {
```

Comment:
Perhaps we could use fastutil's `Int2IntMap`?https://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/Int2IntMap.html